File: app/src/main/java/rogue_opcode/containers/Array.java
// Array.java
// Dynamical array container class
//
// Copyright © 2009-2017 Brigham Toskin
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <http://code.google.com/p/rogue-op/wiki/LICENSE>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode.containers;

import java.util.Iterator;


/**
 * When directly accessing the underlying {@code data} array, indices
 * {@code [0, size)} are valid.
 *
 * @param <E> Generic storage type parameter.
 * @see rogue_opcode.containers.Container
 * @author Brigham Toskin
 */
public class Array<E> extends Container<E> implements Iterator<E>, Iterable<E>
{
	private static final long serialVersionUID = -90938033040808162L;
	int mIteratorIndex;

	// c'tor //

	/** Default constructor preallocates space for up to 32 elements.
	 */
	public Array()
	{
		super(32);
	}

	/** Preallocates space for up to {@code pCapacity} elements.
	 * @param pCapacity number of elements to allocate space for.
	 */
	public Array(int pCapacity)
	{
		super(pCapacity);
	}

	// data access /////////////////////////////////////////////////////////////

	/**
	 * @see rogue_opcode.containers.Container#First()
	 */
	@Override
	public E First()
	{
		validate_nonempty();
		return data[0];
	}

	/**
	 * @see rogue_opcode.containers.Container#Last()
	 */
	@Override
	public E Last()
	{
		validate_nonempty();
		return data[size - 1];
	}

	/**
	 * Retrieve element at index {@code pOffset}.
	 *
	 * @param pIndex element index to retrieve.
	 * @see rogue_opcode.containers.Container#At(int)
	 */
	@Override
	public E At(int pIndex)
	{
		validate_index(pIndex);
		return data[pIndex];
	}

	/** Appends a new element to the end of the array, reallocating the buffer
	 * if necessary.
	 * @param pVal the item to append.
	 */
	public void Append(E pVal)
	{
		resize_inc();
		data[size-1] = pVal;
	}


	@Override
	public boolean hasNext()
	{
		return (mIteratorIndex < size);
	}

	@Override
	public E next()
	{
		validate_index(mIteratorIndex);
		return data[mIteratorIndex ++];
	}

	@Override
	public void remove()
	{
	}

	@Override
	public Iterator<E> iterator()
	{
		mIteratorIndex = 0;
		return this;
	}
}

File: lib/src/main/java/rogue_opcode/containers/Array.kt
// Array.kt
// Dynamical array container class
//
// Copyright © 2009-2017 Brigham Toskin
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <https://github.com/IonoclastBrigham/RogueOp/blob/master/LICENSE.md>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode.containers


/**
 * When directly accessing the underlying `mData` array, indices
 * `[0, mSize)` are valid.
 *
 * @param <E> Generic storage type parameter.
 * @see rogue_opcode.containers.Container
 * @author Brigham Toskin
 */
@Suppress("UNCHECKED_CAST")
open class Array<E> : Container<E>, MutableIterator<E>, Iterable<E> {
	companion object {
		private val serialVersionUID = -90938033040808162L
	}


	internal var mIteratorIndex: Int = 0

	// c'tor //

	/** Default constructor preallocates space for up to 32 elements.
	 */
	constructor() : super(32)

	/**
	 * Preallocates space for up to `pCapacity` elements.
	 * @param pCapacity number of elements to allocate space for.
	 */
	constructor(pCapacity: Int) : super(pCapacity)

	// mData access /////////////////////////////////////////////////////////////

	/**
	 * @see rogue_opcode.containers.Container.First
	 */
	override fun First(): E {
		validate_nonempty()
		return mData[0] as E
	}

	/**
	 * @see rogue_opcode.containers.Container.Last
	 */
	override fun Last(): E {
		validate_nonempty()
		return mData[LastIndex] as E
	}

	/**
	 * Retrieve element at index `pOffset`.

	 * @param pIndex element index to retrieve.
	 * *
	 * @see rogue_opcode.containers.Container.At
	 */
	override fun At(pIndex: Int): E {
		validate_index(pIndex)
		return mData[pIndex] as E
	}

	/** Appends a new element to the end of the array, reallocating the buffer
	 * if necessary.
	 * @param pVal the item to append.
	 */
	open fun Append(pVal: E) {
		resize_inc()
		mData[LastIndex] = pVal
	}

	// Iterators ///////////////////////////////////////////////////////////////

	override fun hasNext(): Boolean {
		return mIteratorIndex < mSize
	}

	override fun next(): E {
		validate_index(mIteratorIndex)
		return mData[mIteratorIndex++] as E
	}

	override fun remove() {} // TODO

	override fun iterator(): MutableIterator<E> {
		mIteratorIndex = 0
		return this
	}
}
